Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,533 @@
+        let userBitpatternAsPointer = UnsafeMutableRawPointer.init(bitPattern: UInt(userbitPattern))
+
+        _debugPrint("io_uring_poll_update fileDescriptor[\(fileDescriptor)] oldPollmask[\(oldPollmask)] newPollmask[\(newPollmask)]  userBitpatternAsPointer[\(String(describing:userBitpatternAsPointer))]")
+
+        self.withSQE { sqe in
+            // "Documentation" for multishot polls and updates here:
+            // https://git.kernel.dk/cgit/linux-block/commit/?h=poll-multiple&id=33021a19e324fb747c2038416753e63fd7cd9266
+            var flags = IORING_POLL_UPDATE_EVENTS | IORING_POLL_UPDATE_USER_DATA
+            if multishot {
+                flags |= IORING_POLL_ADD_MULTI       // ask for multiple updates
+            }
+
+            CNIOLinux.io_uring_prep_poll_update(sqe, bitpatternAsPointer, bitpatternAsPointer, newPollmask, flags)
+            CNIOLinux.io_uring_sqe_set_data(sqe, userBitpatternAsPointer)
+        }
+        
+        if submitNow {
+            self.io_uring_flush()
+        }
+    }
+
+    internal func _debugPrint(_ s : @autoclosure () -> String)
+    {
+        #if SWIFTNIO_IO_URING_DEBUG_URING
+        print("L [\(NIOThread.current)] " + s())
+        #endif
+    }
+
+    // We merge results into fdEvents on (fd, registrationID) for the given CQE
+    // this minimizes amount of events propagating up and allows Selector to discard
+    // events with an old sequence identifier.
+    internal func _process_cqe(events: UnsafeMutablePointer<UringEvent>, cqeIndex: Int, multishot: Bool) {
+        let bitPattern = UInt(bitPattern:io_uring_cqe_get_data(cqes[cqeIndex]))
+        let uringUserData = URingUserData(rawValue: UInt64(bitPattern))
+        let result = cqes[cqeIndex]!.pointee.res
+
+        switch uringUserData.eventType {
+        case .poll:
+            switch result {
+            case -ECANCELED:
+                var pollError : UInt32 = 0
+                assert(uringUserData.fileDescriptor >= 0, "fd must be zero or greater")
+                if multishot { // -ECANCELED for streaming polls, should signal error
+                    pollError = Uring.POLLERR | Uring.POLLHUP
+                } else {       // this just signals that Selector just should resubmit a new fresh poll
+                    pollError = Uring.POLLCANCEL
+                }
+                if let current = fdEvents[fdEventKey(uringUserData.fileDescriptor, uringUserData.registrationID)] {
+                    fdEvents[fdEventKey(uringUserData.fileDescriptor, uringUserData.registrationID)] = current | pollError
+                } else {
+                    fdEvents[fdEventKey(uringUserData.fileDescriptor, uringUserData.registrationID)] = pollError
+                }
+                break
+            case -EINVAL:
+                _debugPrint("Failed poll with -EINVAL for cqeIndex[\(cqeIndex)]")
+                break
+            case -EBADF:
```

Comment:
Is there any case when these errors should not be fatal to our process? That is, is there any circumstance where they don't reflect programmer error?

Reply:
Agree that EINVAL should be fatal, EBADF could happen possibly for a close race (would _probably_ be sloppy coding from user, but we would handle it robustly and seems a bit harsh to terminate then). I've removed EINVAL and changed unexpected errors to fatalError instead as it would be a severe implementation error. EAGAIN/EALREADY can happen during certain modification races and would be ok I think, so those are still handled separately.

Reply:
Right now NIO does not tolerate EBADF from its syscalls under any circumstances: the kind of sloppy coding it implies is highly dangerous and will almost certainly crash later on in the process.

Reply:
The reason I wrote _probably_, was that I had a vague feeling there was a valid reason why this happened - and there is, I just removed the handling to refresh my memory and reran the test suite...

Fundamentally its related to the discussion with the assert above - as we can have a stream of pending requests that are pending towards the kernel - and in this case, the main test thread will perform a close (and we have basically prepared to send the deregistration to the kernel, so a close() system call will be performed, which races with our old pending requests (once again, completely async backend bites us, no implicit synchronisation point for the deregistration). So would need to be robust for EBADF for the close race - it would similarly to the assert either require extremely heavy handed synchronisation when changing registrations visavi the kernel, or simply that we are robust when it happens. So in this particular case, its not really sloppy coding that is the root issue, but rather the true asynchronous nature of the io_uring requests.

```
Test Case 'BootstrapTest.testPreConnectedSocketSetsChannelOptionsBeforeChannelInitializer' started at 2021-04-27 13:17:01.157
...
L [NIOThread(actualName = NIO-ELT-42-#0)] io_uring_prep_poll_add fileDescriptor[13] pollMask[24] bitpatternAsPointer[Optional(0x000100000000000d)] submitNow[false] multishot[true]
S [NIOThread(actualName = NIO-ELT-42-#0)] Re-register old SelectorEventSet(rawValue: 1) new SelectorEventSet(rawValue: 3) uringEventSet [24] reg.uringEventSet [8216]
L [NIOThread(actualName = NIO-ELT-42-#0)] io_uring_poll_update fileDescriptor[13] oldPollmask[24] newPollmask[8216]  userBitpatternAsPointer[Optional(0x000200000000000d)]
S [NIOThread(actualName = NIO-ELT-42-#0)] register interested SelectorEventSet(rawValue: 1) uringEventSet [24] registrationID[SelectorRegistrationID(_rawValue: 1)]
...
S [NIOThread(actualName = NIO-ELT-42-#0)] deregister interested BaseSocket { fd=13 } reg.interested.uringEventSet [8216]
L [NIOThread(actualName = NIO-ELT-42-#0)] io_uring_prep_poll_remove fileDescriptor[13] pollMask[8216] bitpatternAsPointer[Optional(0x000100000000000d)] 
...
L [NIOThread(actualName = NIO-ELT-42-#0)] io_uring_flush io_uring_submit needed [1] submission(s), submitted [1] SQE:s out of [7] possible
L [NIOThread(actualName = NIO-ELT-42-#0)] io_uring_flush io_uring_submit needed [2] submission(s), submitted [6] SQE:s out of [6] possible
L [NIOThread(actualName = NIO-ELT-42-#0)] io_uring_peek_batch_cqe found [12] events
Q [NIOThread(actualName = NIO-ELT-42-#0)] io_uring_peek_batch_cqe CQE:s [0x00007fe988000da0] - ring flags are [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 0 = fd[11] eventType[poll] registrationID[0] res [1] flags [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 1 = fd[12] eventType[pollModify] registrationID[65535] res [0] flags [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 2 = fd[13] eventType[poll] registrationID[0] res [-9] flags [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 3 = fd[13] eventType[pollModify] registrationID[0] res [-2] flags [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 4 = fd[14] eventType[poll] registrationID[1] res [8192] flags [2]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 5 = fd[14] eventType[pollModify] registrationID[1] res [0] flags [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 6 = fd[14] eventType[poll] registrationID[1] res [8192] flags [2]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 7 = fd[14] eventType[pollModify] registrationID[1] res [0] flags [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 8 = fd[14] eventType[poll] registrationID[1] res [8193] flags [2]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 9 = fd[13] eventType[pollDelete] registrationID[0] res [-2] flags [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 10 = fd[12] eventType[poll] registrationID[65535] res [-125] flags [0]
Q [NIOThread(actualName = NIO-ELT-42-#0)] 11 = fd[12] eventType[pollDelete] registrationID[65535] res [0] flags [0]
Fatal error: Failed poll with unexpected error (-9 for cqeIndex[2]: file NIO/LinuxUring.swift, line 360
Current stack trace:
```

Here we can see how all three requests are sent together in this particular circumstance, but are clearly processed out of order and

Reply:
@hassila if there are no ordering guarantees, how are we then guaranteeing that we don't get stuck for example? What if a `pollModify` unregisters say from `.read` meant for `fd=123 (registrationID=1)` but gets processed for `fd=123 (registrationID=2)`?

Reply:
@weissi the registration id is part of the user data that is used to match the request on the kernel side, so we can't modify polls for other registrations.

Reply:
@hassila okay, so we're saying the _only_ thing affected here are requests for fds that are already closed? And I suppose the kernel will _also_ send us `EBADF` for trying to modify `fd=123/regID=1` when in fact we're registered with `fd=123/regID=2`?

Reply:
@weissi also, in this case, the major issue is that the `close()` can race with our pending requests. Later on when moving to a true io_uring implementation, the point will be moot as well, as we would then schedule `read()` directly and not do any modifications (just cancellations of reads, where we can choose to throw away any racy result returned if semantically correct to do so if a user closes a channel while there is inbound data arriving).

Reply:
> @hassila okay, so we're saying the _only_ thing affected here are requests for fds that are already closed? And I suppose the kernel will _also_ send us `EBADF` for trying to modify `fd=123/regID=1` when in fact we're registered with `fd=123/regID=2`?

@weissi It would give us an ENOENT for that case, but an error - yes.

Reply:
Thanks for the clarifications. This all makes sense. I think in this particular case accepting `EBADF` is then unfortunately reasonable. I think we should add a good comment for that. The key thing obviously is that there is no synchronisation barrier with the kernel so we have no idea when the kernel will have apply our submission requests.
@Lukasa wdyt?

Reply:
I'm adding some comments now.

Reply:
I'm in principle fine with this, but I'd like to be confident that iouring absolutely does take user data into account when identifying a poll request.

I can't find a specific citation that implies that the IORING_OP_POLL_ADD or IORING_OP_POLL_REMOVE events do anything in particular with the user data other than return it. That doesn't mean they don't, only that I haven't yet found any evidence that they do.

The thing that worries me is the question of when the kernel resolves the file descriptor into the file description. I am entirely confident that, once the kernel has done that resolution, closing the userland fd will not cause us any difficulty. What gives me pause is the possibility that we close one fd and then obtain it from a new open, registering it on a new uring, all before the first uring processes the SQE referring to the old (now closed-and-replaced) file descriptor. What happens here?

My gut instinct is that io_uring does not use the user_data to distinguish these requests and so can return a spurious poll completion. Does that sound right?

Reply:
AFAIU, looking [at the implementation](https://git.kernel.dk/cgit/linux-block/diff/fs/io_uring.c?h=for-5.13%2Fio_uring&id=9ba5fac8cf3b607652397f863dc229bbc8c3cbc1&context=40&ignorews=0&dt=0)
`io_poll_find()` is used to find the request for both poll update/cancel SQE:s and there we have the check `if (sqe_addr != req->user_data)` matching on user_data. Poll add doesn't need to do more than return it. 

> What gives me pause is the possibility that we close one fd and then obtain it from a new open, registering it on a new uring, all before the first uring processes the SQE referring to the old (now closed-and-replaced) file descriptor. What happens here?

I'll see if I can ask @isilence or @axboe to comment on that specific aspect.

Reply:
> AFAIU, looking [at the implementation](https://git.kernel.dk/cgit/linux-block/diff/fs/io_uring.c?h=for-5.13%2Fio_uring&id=9ba5fac8cf3b607652397f863dc229bbc8c3cbc1&context=40&ignorews=0&dt=0)
> `io_poll_find()` is used to find the request for both poll update/cancel SQE:s and there we have the check `if (sqe_addr != req->user_data)` matching on user_data. Poll add doesn't need to do more than return it.

Right, it only does matching against it. Also, modify/remove don't take an fd, so matching uses only that passed data.

> > What gives me pause is the possibility that we close one fd and then obtain it from a new open, registering it on a new uring, all before the first uring processes the SQE referring to the old (now closed-and-replaced) file descriptor. What happens here?

fds are resolved during submission, after that the kernel operates with referenced `struct file *`, so no update of userspace file tables can affect it. So, not non-sqpoll rings it's stable right after `io_uring_submit()` and so the corresponding enter syscall returns.

sqpoll rings may lag in that sense because the submission is done by async process that should wake up and reap SQEs. But SQ can be waited upon if needed. IIRC there was a helper in liburing.

Reply:
> > AFAIU, looking [at the implementation](https://git.kernel.dk/cgit/linux-block/diff/fs/io_uring.c?h=for-5.13%2Fio_uring&id=9ba5fac8cf3b607652397f863dc229bbc8c3cbc1&context=40&ignorews=0&dt=0)
> > `io_poll_find()` is used to find the request for both poll update/cancel SQE:s and there we have the check `if (sqe_addr != req->user_data)` matching on user_data. Poll add doesn't need to do more than return it.
> 
> Right, it only does matching against it. Also, modify/remove don't take an fd, so matching uses only that passed in user_data/sqe_addr.
> 
> > > What gives me pause is the possibility that we close one fd and then obtain it from a new open, registering it on a new uring, all before the first uring processes the SQE referring to the old (now closed-and-replaced) file descriptor. What happens here?
> 
> fds are resolved during submission, after that the kernel operates with referenced `struct file *`, so no update of userspace file tables can affect it. So, not non-sqpoll rings it's stable right after `io_uring_submit()` and so the corresponding enter syscall returns.
> 
> sqpoll rings may lag in that sense because the submission is done by async process that should wake up and reap SQEs. But SQ can be waited upon if needed. IIRC there was a helper in liburing.

Reply:
Thank you very much @isilence for clarification. We currently don't support SQPOLL rings for Swift NIO so given that the kernel operates with the [`struct file *`](https://www.oreilly.com/library/view/linux-device-drivers/0596000081/ch03s04.html) after the submit, we should not get updates just because someone else happened to get the same fd in userspace after close/open. Must admit it is always interesting to straddle and peek into kernel land ;-)

Reply:
Great, ok. In that case I think I'm happy with the code as-is, so long as we comment about this (and ideally reference this thread).

Reply:
Thanks for chiming in @isilence, much appreciated.

Reply:
fwiw, `struct file` may be better known for some as "open file description" or OFD

Reply:
Ok, I've updated the comment and put in a reference to this discussion too.